pennylane/_qubit_device.py

```diff
@@ -688,3 +697,96 @@ def sample(self, observable):
         unraveled_indices = [2] * len(device_wires)
         indices = np.ravel_multi_index(samples.T, unraveled_indices)
         return observable.eigvals[indices]
+
+    def adjoint_jacobian(self, tape):
+        """Implements the adjoint method outlined in
+        `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+        After a forward pass, the circuit is reversed by iteratively applying inverse (adjoint)
+        gates to scan backwards through the circuit. This method is similar to the reversible
+        method, but has a lower time overhead and a similar memory overhead.
+
+        .. note::
+            The adjoint differentation method has the following restrictions:
+
+            * As it requires knowledge of the statevector, only statevector simulator devices can be
+              used.
+
+            * Only expectation values are supported as measurements.
+
+        Args:
+            tape (.QuantumTape): circuit that the function takes the gradient of
+
+        Returns:
+            array: the derivative of the tape with respect to trainable parameters.
+            Dimensions are ``(len(observables), len(trainable_params))``.
+
+        Raises:
+            QuantumFunctionError: if the input tape has measurements that are not expectation values
+                or contains a multi-parameter operation aside from :class:`~.Rot`
+        """
+
+        for m in tape.measurements:
+            if m.return_type is not qml.operation.Expectation:
+                raise qml.QuantumFunctionError(
+                    "Adjoint differentiation method does not support"
+                    f" measurement {m.return_type.value}"
+                )
+
+            if not hasattr(m.obs, "base_name"):
+                m.obs.base_name = None  # This is needed for when the observable is a tensor product
```

Comment:
:thinking: it's an `Operation` property:
```python
    @property
    def base_name(self):
        """Get base name of the operator."""
        return self.__class__.__name__
```
Line 736-737 is needed since we use the `base_name` in `default.qubit` [here](https://github.com/PennyLaneAI/pennylane/blob/86727f5d1389fee4bcc247160d3ed99406898860/pennylane/devices/default_qubit.py#L200), and `base_name` is not defined for tensor products.

Reply:
`name` gets modified if the function is inverted.

Maybe instead of these lines, in `default.qubit` we can use `get_attr(operation, base_name, None)`

Reply:
Oh, weird 🤔 This property skipped my attention. I have a feeling it got added during the circuit drawer PR.

Reply:
I remember during my technical interview I made the change in `default.qubit` so inverted operations could still call the special functions (ie `_apply_x`) instead of always using the more general `_apply_unitary`.  

`base_name` an operation property at the time.